### PR TITLE
imgproc: refactor EMD to reduce C-API usage

### PR DESCRIPTION
**Pull request by mshabunin**
- added more tests for EMD
- refactored to remove CvArr
- used BufferArea for memory allocations
- renamed functions and variables and formatted the code
- kept legacy functions intact in separate header

### CODE REVIEW

File: modules/imgproc/src/emd_new.cpp [hunk 1]
```diff
@@ -0,0 +1,1011 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+/*
+    Partially based on Yossi Rubner code:
```
[the remark this answers is not in the document]
Comment (mshabunin) [reply modules/imgproc/src/emd_new.cpp:6]: No, all code and license are from old `emd.cpp`.

File: modules/imgproc/test/test_emd.cpp [hunk 2]
```diff
@@ -1,93 +1,196 @@
...
+                const float emd0 = calcEMD(Mat(w3), Mat(w4), flow, dt, 2);
+                EXPECT_NEAR(emd0, emd, 1e-6);
+            }
+        }
+        const vector<DistanceTypes> bad_types {DIST_L12, DIST_FAIR, DIST_WELSCH, DIST_HUBER};
+        for (const auto& dt : bad_types)
+        {
+            SCOPED_TRACE(cv::format("dt=%d", dt));
+            EXPECT_THROW(EMD(w1, w2, dt), cv::Exception);
+        }
```
Comment (asmorkalov) [inline modules/imgproc/test/test_emd.cpp:143]: It'll be dedicated test with parameter.

Comment (mshabunin) [reply modules/imgproc/test/test_emd.cpp]: I decided to extract this block to separate _invalid_ test. It seem to be too small to be in a separate parametrized test. `SCOPED_TRACE` would help with debugging test failures.

**Review by asmorkalov — APPROVED** [review]

**Comment by opencv-alalek** [comment]
There are regressions in 32-bit builders:
- https://pullrequest.opencv.org/buildbot/builders/4_x_noOCL-lin32/builds/100626
- https://pullrequest.opencv.org/buildbot/builders/4_x_noICV-win32-vc14/builds/100072

**Pull request merged**